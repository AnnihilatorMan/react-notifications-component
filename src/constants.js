export const NOTIFICATION_BASE_CLASS = "notification-item";

export const CONTAINER = {
  BOTTOM_LEFT: "bottom-left",
  BOTTOM_RIGHT: "bottom-right",
  TOP_LEFT: "top-left",
  TOP_RIGHT: "top-right"
};

export const INSERTION = {
  TOP: "top",
  BOTTOM: "bottom"
};

export const NOTIFICATION_TYPE = {
  SUCCESS: "success",
  DANGER: "danger",
  INFO: "info",
  DEFAULT: "default",
  WARNING: "warning"
};

export const NOTIFICATION_STAGE = {
  // used for both sliding and animation at the same time
  SLIDING_ANIMATION_EXIT: "SLIDING_ANIMATION_EXIT",

  // used for onTouchMove/onTouchEnd to slide and animate
  TOUCH_SLIDING_ANIMATION_EXIT: "TOUCH_SLIDING_ANIMATION_EXIT",

  // used by API call to remove notification
  REMOVAL: "REMOVAL"
};
