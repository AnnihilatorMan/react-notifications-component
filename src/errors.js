export default {
  // dismiss icon option
  DISMISS_ICON_CLASS: "className property of dismissIcon option is required",
  DISMISS_ICON_CONTENT: "content property of dismissIcon option is required",
  DISMISS_ICON_STRING: "className property of dismissIcon option must be a String",
  DISMISS_ICON_INVALID: "content property of dismissIcon option must be a valid React element",

  // animations
  ANIMATION_IN: "animationIn option must be an array",
  ANIMATION_OUT: "animationOut option must be an array",

  // dismiss
  DISMISS_REQUIRED: "duration property of dismiss option is required",
  DISMISS_NUMBER: "duration property of dismiss option must be a Number",
  DISMISS_POSITIVE: "duration property of dismiss option must be a positive Number",

  // title
  TITLE_STRING: "title option must be a String.",

  // message
  MESSAGE_REQUIRED: "message option is required",
  MESSAGE_STRING: "message option must be a String",

  // type
  TYPE_REQUIRED: "type option is required",
  TYPE_STRING: "type option must be a String",
  TYPE_NOT_EXISTENT: "type option not found",

  // container
  CONTAINER_REQUIRED: "container option is required",
  CONTAINER_STRING: "container option must be a String",

  // dismissable
  DISMISSABLE_CLICK_BOOL: "click property of dismissable option must be a Boolean",
  DISMISSABLE_TOUCH_BOOL: "touch property of dismissable option must be a Boolean",

  // width
  WIDTH_NUMBER: "width option must be a Number",

  // insert
  INSERT_STRING: "insert option must be a String",

  // transition
  TRANSITION_DURATION_NUMBER: "duration property of transition option must be a Number",
  TRANSITION_CUBICBEZIER_NUMBER: "cubicBezier property of transition option must be a String",
  TRANSITION_DELAY_NUMBER: "delay property of transition option must be a Number",

  // custom types
  TYPE_NOT_FOUND: "custom type not found"
};
