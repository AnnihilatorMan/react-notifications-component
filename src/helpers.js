import React from "react";
import {
  handleTouchSlidingAnimationExit,
  handleSlidingAnimationExit
} from "./stage-helpers";

import {
  NOTIFICATION_BASE_CLASS,
  CONTAINER,
  INSERTION,
  NOTIFICATION_TYPE,
  NOTIFICATION_STAGE
} from "./constants";

import {
  isNullOrUndefined,
  isString,
  isNumber,
  isBoolean,
  isArray
} from "./utils";

import ERROR from "./errors";

export function isBottomContainer(container) {
  return (
    container === CONTAINER.BOTTOM_LEFT ||
    container === CONTAINER.BOTTOM_RIGHT
  );
}

export function isTopContainer(container) {
  return (
    container === CONTAINER.TOP_LEFT ||
    container === CONTAINER.TOP_RIGHT
  );
}

export function shouldNotificationHaveSliding(notification) {
  return (
    // slide DOWN if container is top and insertion is at top
    (notification.insert === INSERTION.TOP && isTopContainer(notification.container)) ||
    // slide UP if container is bottom and insertion is at bottom
    (notification.insert === INSERTION.BOTTOM && isBottomContainer(notification.container))
  );
}

export function cssWidth(width) {
  return width ? `${width}px` : undefined;
}

function htmlClassesForExistingType(type) {
  switch (type.toLowerCase()) {
    case NOTIFICATION_TYPE.DEFAULT:
      return [NOTIFICATION_BASE_CLASS, "notification-default"];
    case NOTIFICATION_TYPE.SUCCESS:
      return [NOTIFICATION_BASE_CLASS, "notification-success"];
    case NOTIFICATION_TYPE.DANGER:
      return [NOTIFICATION_BASE_CLASS, "notification-danger"];
    case NOTIFICATION_TYPE.WARNING:
      return [NOTIFICATION_BASE_CLASS, "notification-warning"];
    case NOTIFICATION_TYPE.INFO:
      return [NOTIFICATION_BASE_CLASS, "notification-info"];
    default:
      throw new Error(`Type '${type}' is not valid nor user-defined`);
  }
}

export function getHtmlClassesForType(notification) {
  const {
    // option set by user
    type,
    // array value containing all custom types
    userDefinedTypes
  } = notification;

  if (notification.content) {
    // return only base class if type is not defined
    return [NOTIFICATION_BASE_CLASS];
  } else if (isNullOrUndefined(userDefinedTypes)) {
    // existing type
    return htmlClassesForExistingType(type);
  }

  // look for custom type if any defined
  const foundType = userDefinedTypes.find(q => q.name === type);

  if (!foundType) {
    // custom type not found, throw error
    throw new Error(ERROR.TYPE_NOT_FOUND);
  }

  // append base class to html classes
  return [NOTIFICATION_BASE_CLASS].concat(foundType.htmlClasses);
}

export function getNotificationsForMobileView(notifications) {
  const {
    TOP_LEFT,
    TOP_RIGHT,
    BOTTOM_LEFT,
    BOTTOM_RIGHT
  } = CONTAINER;

  const bottom = [];
  const top = [];

  notifications.forEach((notification) => {
    const container = notification.container.toLowerCase();
    switch (container) {
      case TOP_LEFT:
      case TOP_RIGHT:
        top.push(notification);
        break;
      case BOTTOM_LEFT:
      case BOTTOM_RIGHT:
        bottom.push(notification);
        break;
      default:
        // throw error just in case validation didn't work properly
        throw new Error(`Container ${notification.container} is not valid`);
    }
  });

  return { top, bottom };
}

export function getNotificationsForEachContainer(notifications) {
  const topLeft = [];
  const topRight = [];
  const bottomLeft = [];
  const bottomRight = [];

  notifications.forEach((notification) => {
    const {
      TOP_LEFT,
      TOP_RIGHT,
      BOTTOM_LEFT,
      BOTTOM_RIGHT
    } = CONTAINER;

    const container = notification.container.toLowerCase();
    switch (container) {
      case TOP_LEFT:
        topLeft.push(notification);
        break;
      case TOP_RIGHT:
        topRight.push(notification);
        break;
      case BOTTOM_LEFT:
        bottomLeft.push(notification);
        break;
      case BOTTOM_RIGHT:
        bottomRight.push(notification);
        break;
      default:
        // throw error just in case validation didn't work properly
        throw new Error(`Container ${notification.container} is not valid`);
    }
  });

  return {
    topLeft,
    topRight,
    bottomLeft,
    bottomRight
  };
}

export function getInitialSlidingState({ notification, isFirstNotification }) {
  // no sliding needed for first notification in container
  const hasSliding = shouldNotificationHaveSliding(notification) && !isFirstNotification;
  const state = {};

  // set default classes for animated element
  state.animatedElementClasses = getHtmlClassesForType(notification);
  state.rootElementStyle = {
    height: "0",
    marginBottom: 0,
    overflow: "hidden",
    width: cssWidth(notification.width)
  };

  if (hasSliding) {
    // hide content by toggling visibility while sliding
    state.animatedElementClasses.push("invisible");
  } else if (notification.animationIn && notification.animationIn.length > 0) {
    // add user defined notification classes if sliding will not occur
    notification.animationIn.forEach(item => state.animatedElementClasses.push(item));
  }

  state.hasSliding = hasSliding;
  return state;
}

export function handleStageTransition(notification, state) {
  let animatedElementClasses;
  let rootElementStyle;

  const {
    animatedElementClasses: stateAnimatedElementClasses,
    rootElementStyle: stateRootStyle,
    currentX,
    startX
  } = state;

  if (notification.stage === NOTIFICATION_STAGE.TOUCH_SLIDING_ANIMATION_EXIT) {
    return handleTouchSlidingAnimationExit(notification, currentX, startX);
  } else if (notification.stage === NOTIFICATION_STAGE.SLIDING_ANIMATION_EXIT) {
    return handleSlidingAnimationExit(notification);
  } else if (notification.resized) {
    // window got resized, do not apply animations
    rootElementStyle = stateRootStyle;
    animatedElementClasses = getHtmlClassesForType(notification);
  } else {
    // use values from state
    rootElementStyle = stateRootStyle;
    animatedElementClasses = stateAnimatedElementClasses;
  }

  return {
    rootElementStyle,
    animatedElementClasses
  };
}

export function getCubicBezierTransition(
  duration = 500,
  cubicBezier = "linear",
  delay = 0,
  property = "height"
) {
  return `${duration}ms ${property} ${cubicBezier} ${delay}ms`;
}

export function hasFullySwiped(diffX) {
  const swipeLength = Math.abs(diffX);
  const requiredSwipeLength = (40 / 100) * window.innerWidth;

  return swipeLength >= requiredSwipeLength;
}

export function getRootHeightStyle(notification, scrollHeight) {
  return {
    height: `${scrollHeight}px`,
    width: cssWidth(notification.width),
    transition: getCubicBezierTransition(
      notification.slidingExit.duration,
      notification.slidingExit.cubicBezier,
      notification.slidingExit.delay
    )
  };
}

export function getIconHtmlContent(notification, onClickHandler) {
  // use icon defined by user
  if (notification.dismissIcon) {
    return (
      <div
        className={notification.dismissIcon.className}
        onClick={onClickHandler}
      >
        {notification.icon}
      </div>
    );
  }

  // use standard icon
  return (
    <div
      className="notification-close"
      onClick={onClickHandler}
    >
      <span>&times;</span>
    </div>
  );
}

export function validateDismissIconOption(dismissIcon) {
  if (isNullOrUndefined(dismissIcon)) return;

  const { className, content } = dismissIcon;

  if (!className) {
    // className is required
    throw new Error(ERROR.DISMISS_ICON_CLASS);
  } else if (!isString(className)) {
    // must be an array value
    throw new Error(ERROR.DISMISS_ICON_STRING);
  } else if (!content) {
    // content is required
    throw new Error(ERROR.DISMISS_ICON_CONTENT);
  } else if (!React.isValidElement(content)) {
    // must be a valid react element
    throw new Error(ERROR.DISMISS_ICON_INVALID);
  }
}

export function validateAnimationIn(animationIn) {
  if (isNullOrUndefined(animationIn)) {
    return [];
  } if (!isArray(animationIn)) {
    // must be an array
    throw new Error(ERROR.ANIMATION_IN);
  }

  return animationIn;
}

export function validateAnimationOut(animationOut) {
  if (isNullOrUndefined(animationOut)) {
    return [];
  } else if (!isArray(animationOut)) {
    // must be an array
    throw new Error(ERROR.ANIMATION_OUT);
  }

  return animationOut;
}

export function validateTimeoutDismissOption(dismiss) {
  // skip if option is not defined
  if (isNullOrUndefined(dismiss)) return;

  if (isNullOrUndefined(dismiss.duration)) {
    throw new Error(ERROR.DISMISS_REQUIRED);
  } else if (!isNumber(dismiss.duration)) {
    // must be a number
    throw new Error(ERROR.DISMISS_NUMBER);
  } else if (dismiss.duration < 0) {
    // duration must be positive
    throw new Error(ERROR.DISMISS_POSITIVE);
  }
}

export function validateTransition(transition, defaults) {
  const { duration, cubicBezier, delay } = defaults;
  const transitionOptions = transition || {};

  if (isNullOrUndefined(transitionOptions.duration)) {
    // set default duration
    transitionOptions.duration = duration;
  }

  if (isNullOrUndefined(transitionOptions.cubicBezier)) {
    // set default timing function
    transitionOptions.cubicBezier = cubicBezier;
  }

  if (isNullOrUndefined(transitionOptions.delay)) {
    // set default delay
    transitionOptions.delay = delay;
  }

  if (!isNumber(transitionOptions.duration)) {
    // throw if duration is NaN
    throw new Error(ERROR.TRANSITION_DURATION_NUMBER);
  } else if (!isString(transitionOptions.cubicBezier)) {
    // throw if cubicBezier is not a String
    throw new Error(ERROR.TRANSITION_CUBICBEZIER_NUMBER);
  } else if (!isNumber(transitionOptions.delay)) {
    // throw if duration is NaN
    throw new Error(ERROR.TRANSITION_DELAY_NUMBER);
  }

  return transitionOptions;
}

export function validateTitle(notification) {
  const { content, title } = notification;

  // option not required if content is defined
  if (content) return;

  if (title === null || title === undefined) {
    // title is required
    throw new Error(ERROR.TITLE_REQUIRED);
  } else if (!isString(title)) {
    // title must be a String
    throw new Error(ERROR.TITLE_STRING);
  }
}

export function validateMessage(notification) {
  const { content, message } = notification;

  // option not required if content is defined
  if (content) return;

  if (message === null || message === undefined) {
    throw new Error(ERROR.MESSAGE_REQUIRED);
  } else if (!isString(message)) {
    throw new Error(ERROR.MESSAGE_STRING);
  }
}

export function validateType(notification, userDefinedTypes) {
  const { content, type } = notification;

  if (content) {
    // option not required if content is defined
    return undefined;
  } else if (isNullOrUndefined(type)) {
    // type is required
    throw new Error(ERROR.TYPE_REQUIRED);
  } else if (!isString(type)) {
    // type must be a String
    throw new Error(ERROR.TYPE_STRING);
  } else if (
    isNullOrUndefined(userDefinedTypes) &&
    type !== NOTIFICATION_TYPE.SUCCESS &&
    type !== NOTIFICATION_TYPE.DANGER &&
    type !== NOTIFICATION_TYPE.INFO &&
    type !== NOTIFICATION_TYPE.DEFAULT &&
    type !== NOTIFICATION_TYPE.WARNING
  ) {
    throw new Error(ERROR.TYPE_NOT_EXISTENT);
  }

  return type.toLowerCase();
}

export function validateContainer(container) {
  if (isNullOrUndefined(container)) {
    // container is required
    throw new Error(ERROR.CONTAINER_REQUIRED);
  } else if (!isString(container)) {
    // container must be a String
    throw new Error(ERROR.CONTAINER_STRING);
  }

  return container.toLowerCase();
}

export function validateDismissable(dismissable) {
  if (isNullOrUndefined(dismissable)) {
    return {
      // set dissmisable by click as default
      click: true,
      // set dismissable on swipe as default
      touch: true
    };
  }

  const option = dismissable;

  if (isNullOrUndefined(option.click)) {
    // set dissmisable by click as default
    option.click = true;
  }

  if (isNullOrUndefined(option.touch)) {
    // set dismissable on swipe as default
    option.touch = true;
  }

  if (!isBoolean(option.click)) {
    // must be boolean
    throw new Error(ERROR.DISMISSABLE_CLICK_BOOL);
  } else if (!isBoolean(option.touch)) {
    // must be boolean
    throw new Error(ERROR.DISMISSABLE_TOUCH_BOOL);
  }

  return option;
}

export function validateInsert(insert) {
  if (isNullOrUndefined(insert)) {
    // default to top if not defined
    return "top";
  } else if (!isString(insert)) {
    // must be a string value (top|bottom)
    throw new Error(ERROR.INSERT_STRING);
  }

  return insert;
}

export function validateWidth(width) {
  if (!isNumber(width)) {
    // width must be a valid Number
    throw new Error(ERROR.WIDTH_NUMBER);
  }

  return width;
}

export function validateUserDefinedTypes(notification, definedTypes) {
  const { content, type } = notification;

  // no need to validate type if content is defined
  if (content) return undefined;

  if (
    type === NOTIFICATION_TYPE.SUCCESS ||
    type === NOTIFICATION_TYPE.DANGER ||
    type === NOTIFICATION_TYPE.INFO ||
    type === NOTIFICATION_TYPE.DEFAULT ||
    type === NOTIFICATION_TYPE.WARNING ||
    isNullOrUndefined(definedTypes)
  ) return undefined;

  // search for custom type in array
  if (!definedTypes.find(p => p.name === type)) {
    // custom type not found, throw exception
    throw new Error(ERROR.TYPE_NOT_FOUND);
  }

  return definedTypes;
}

function getRandomId() {
  return Math.random();
}

export function getNotificationOptions(options, userDefinedTypes) {
  const notification = options;
  const {
    insert,
    container,
    dismissIcon,
    animationIn,
    animationOut,
    slidingEnter,
    slidingExit,
    touchSlidingBack,
    touchSlidingExit,
    dismissable,
    dismiss,
    width
  } = notification;

  // for now we'll use Math.random for id
  notification.id = getRandomId();

  // validate notification's title
  validateTitle(notification);

  // validate notification's body message
  validateMessage(notification);

  // validate notification's type
  notification.type = validateType(notification, userDefinedTypes);

  if (userDefinedTypes && !notification.content) {
    // validate user defined types if any
    notification.userDefinedTypes = validateUserDefinedTypes(notification, userDefinedTypes);
  }

  // validate notification's container
  notification.container = validateContainer(container);

  // set default insertion
  notification.insert = validateInsert(insert);

  // set default dismissable options
  notification.dismissable = validateDismissable(dismissable);

  // assert `dismissIcon` matches rules
  validateDismissIconOption(dismissIcon);

  // validate `animationIn` array
  notification.animationIn = validateAnimationIn(animationIn);

  // validate `animationOut` array
  notification.animationOut = validateAnimationOut(animationOut);

  if (!isNullOrUndefined(width)) {
    // set width option only if defined
    notification.width = validateWidth(width);
  }

  // defaults
  const slidingEnterDefaults = { duration: 600, cubicBezier: "linear", delay: 0 };
  const slidingExitDefaults = { duration: 600, cubicBezier: "linear", delay: 0 };
  const swipeBackDefaults = { duration: 600, cubicBezier: "ease-in", delay: 0 };
  const swipeExitDefaults = { duration: 600, cubicBezier: "ease-in", delay: 0 };
  const swipeCompleteDefaults = { duration: 300, cubicBezier: "ease-in", delay: 0 };

  // set defaults for sliding enter transition
  notification.slidingEnter = validateTransition(slidingEnter, slidingEnterDefaults);

  // set defaults for sliding exit transition
  notification.slidingExit = validateTransition(slidingExit, slidingExitDefaults);

  // set defaults for sliding back on touchEnd/touchCancel
  notification.touchSlidingBack = validateTransition(touchSlidingBack, swipeBackDefaults);

  // let it empty object if undefined
  notification.touchSlidingExit = notification.touchSlidingExit || {};

  // set defaults for swipe transition when swipe is complete
  notification.touchSlidingExit.swipe = validateTransition(
    touchSlidingExit.swipe || {},
    swipeExitDefaults
  );

  // set defaults for fade transition when swipe is complete
  notification.touchSlidingExit.fade = validateTransition(
    touchSlidingExit.fade || {},
    swipeCompleteDefaults
  );

  // assert `dismiss` matches rules
  validateTimeoutDismissOption(dismiss);

  return notification;
}
