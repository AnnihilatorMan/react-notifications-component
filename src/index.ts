import { Container } from 'src/components/Container'
import Store from 'src/store'
export default Container

import {
  iContainerProps,
  iContainerState,
  iNotification,
  NotificationTitleMessage,
  NotificationContent,
  iTransition,
  iTouchTransition,
  iNotificationDismiss,
  iNotificationCustomType,
  iNotificationProps,
  iNotificationState,
  iNotificationParentStyle,
  iStore,
  iNotificationStoreParams,
  NOTIFICATION_CONTAINER,
  NOTIFICATION_TYPE,
  NOTIFICATION_INSERTION,
  NOTIFICATION_REMOVAL_SOURCE
} from './types'

export { Store as store } // otherwise it's breaking change
export { Container as ReactNotifications }

export {
  iContainerProps,
  iContainerState,
  iNotification,
  NotificationTitleMessage,
  NotificationContent,
  iTransition,
  iTouchTransition,
  iNotificationDismiss,
  iNotificationCustomType,
  iNotificationProps,
  iNotificationState,
  iNotificationParentStyle,
  iStore,
  iNotificationStoreParams,
  NOTIFICATION_CONTAINER,
  NOTIFICATION_TYPE,
  NOTIFICATION_INSERTION,
  NOTIFICATION_REMOVAL_SOURCE
}
