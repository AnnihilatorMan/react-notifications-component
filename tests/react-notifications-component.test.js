import React from "react";
import ReactNotificationComponent from "src/react-notification-component";
import Enzyme, { mount } from "enzyme";
import React16Adapter from "enzyme-adapter-react-16";
import sinon from "sinon";
import notificationMock from "tests/mocks/notification.mock";
import { NOTIFICATION_STAGE, BREAKPOINT } from "src/constants";

Enzyme.configure({
  // use react-16 adapter
  adapter: new React16Adapter()
});

describe("Wrapper component", () => {
  let component;
  let clock;

  // helpers
  const state = () => component.state();
  const instance = () => component.instance();
  
  const getTypesMock = () => [{
    name: "awesome",
    htmlClasses: ["awesome"]
  }];

  const getNotificationMock = (edits = {}) => {
    return Object.assign({}, notificationMock, edits);
  };

  const addNotification = (mock) => {
    return instance().addNotification(mock);
  };

  beforeEach(() => {
    component = mount(<ReactNotificationComponent />);

    // set fake timer
    clock = sinon.useFakeTimers();
  });

  afterEach(() => {
    // restore fake timer
    clock.restore();

    // unmount
    component.unmount();
  });

  it("toggleTimeoutRemoval does not change stage if id does not match", () => {
    component = mount(<ReactNotificationComponent />);
    addNotification(getNotificationMock());

    const stage = state().notifications[0].stage;

    // toggle removal
    instance().toggleTimeoutRemoval({ id: 0 });

    // expect stage to match old stage
    expect(state().notifications[0].stage).toEqual(stage);
  });

  it("toggleTouchEnd does not change stage if id does not match", () => {
    component = mount(<ReactNotificationComponent />);
    addNotification(getNotificationMock());

    const stage = state().notifications[0].stage;

    // toggle removal
    instance().toggleTouchEnd({ id: 0 });

    // expect stage to match old stage
    expect(state().notifications[0].stage).toEqual(stage);
  });

  it("removeNotification does not change stage if id does not match", () => {
    component = mount(<ReactNotificationComponent />);
    addNotification(getNotificationMock());

    const stage = state().notifications[0].stage;

    instance().removeNotification({ id: 0 });

    // expect stage to match old stage
    expect(state().notifications[0].stage).toEqual(stage);

    // wait for requestAnimationFrame
    clock.tick(400);

    // expect stage to match old stage
    expect(state().notifications[0].stage).toEqual(stage);
  });

  it("onNotificationClick does not change stage if id does not match", () => {
    component = mount(<ReactNotificationComponent />);
    addNotification(getNotificationMock());

    const stage = state().notifications[0].stage;

    // toggle notification click
    instance().onNotificationClick({ dismissable: { click: true } });

    clock.tick(400);

    // expect stage to match old stage
    expect(state().notifications[0].stage).toEqual(stage);
  });

  it("addNotification adds notifications at the bottom", () => {
    component = mount(<ReactNotificationComponent />);
    const id = addNotification(getNotificationMock({ insert: "BOTTOM" }));
    const id2 = addNotification(getNotificationMock({ insert: "BOTTOM" }));

    expect(state().notifications[0].id).toEqual(id);
    expect(state().notifications[1].id).toEqual(id2);
  });

  it("addNotification adds notifications at the top", () => {
    component = mount(<ReactNotificationComponent />);
    const id = addNotification(getNotificationMock({ insert: "TOP" }));
    const id2 = addNotification(getNotificationMock({ insert: "TOP" }));

    expect(state().notifications[0].id).toEqual(id2);
    expect(state().notifications[1].id).toEqual(id);
  });

  it("component sets state to default values on initialization", () => {
    component = mount(<ReactNotificationComponent />);
    expect(state()).toMatchSnapshot();
  });

  it("component does not set state to default values on initialization", () => {
    component = mount(<ReactNotificationComponent breakpoint={320} isMobile={false} />);
    expect(state()).toMatchSnapshot();
  });

  it("component sets userDefinedTypes on initialization", () => {
    component = mount(<ReactNotificationComponent types={getTypesMock()} />);
    expect(state()).toMatchSnapshot();
  });

  it("handleResize updates width and notifications", () => {
    // mount component
    component = mount(<ReactNotificationComponent />);

    // add notifications
    addNotification(getNotificationMock());
    addNotification(getNotificationMock());
    addNotification(getNotificationMock());

    // set window width to 100px
    window.innerWidth = 100;
    // manually call resize handler
    instance().handleResize();

    // expect width to be set to 100px
    expect(state().width).toBe(100);
    // expect each notification to have resized flag set
    state().notifications.forEach(notification => expect(notification.resized).toBe(true));
  });

  it("removes notification manually", () => {
    let notification;

    // mount
    component = mount(<ReactNotificationComponent />);

    // add notification and store id
    let id = addNotification(getNotificationMock());

    // manually remove notification
    instance().removeNotification(id);

    notification = state().notifications.find(item => item.id === id);

    // expect REMOVAL stage
    expect(notification.stage).toBe(NOTIFICATION_STAGE.MANUAL_REMOVAL);

    // tick
    clock.tick(100);

    notification = state().notifications.find(item => item.id === id);

    // expect SLIDING_ANIMATION_EXIT stage
    expect(notification.stage).toBe(NOTIFICATION_STAGE.SLIDING_ANIMATION_EXIT);
  });

  it("updates stage to SLIDING_ANIMATION_EXIT on toggle-timeout", () => {
    // mount
    component = mount(<ReactNotificationComponent />);

    // add notification
    let id = addNotification(getNotificationMock());

    // call `toggleTimeoutRemoval`
    instance().toggleTimeoutRemoval({ id });

    // get notification from state
    let notifications = component.state("notifications");
    let item = notifications.find(notif => notif.id === id);

    // expect item to be in state
    expect(item).not.toBeNull();
    expect(item).not.toBeUndefined();

    // expect stage to be set to SLIDING_ANIMATION_EXIT
    expect(item.stage).toBe(NOTIFICATION_STAGE.SLIDING_ANIMATION_EXIT);
  });

  it("updates stage to SLIDING_ANIMATION_EXIT on click", () => {
    // mount
    component = mount(<ReactNotificationComponent />);

    // add notification
    let id = addNotification(getNotificationMock());
    const dismissable = { click: true };

    // trigger click
    instance().onNotificationClick({ id, dismissable });

    // tick
    clock.tick(100);

    // find notification by id
    let item = state().notifications.find(elem => elem.id === id);

    // expect stage to be set
    expect(item.stage).toBe(NOTIFICATION_STAGE.SLIDING_ANIMATION_EXIT);
  });

  it("does not update stage to SLIDING_ANIMATION_EXIT on click", () => {
    // mount
    component = mount(<ReactNotificationComponent />);

    // add notification
    const id = addNotification(getNotificationMock());
    const dismissable = {};

    // trigger click
    instance().onNotificationClick({ id, dismissable });

    // tick
    clock.tick(100);

    // find by id
    let item = state().notifications.find(elem => elem.id === id);

    // expect stage to be set
    expect(item.stage).not.toBe(NOTIFICATION_STAGE.SLIDING_ANIMATION_EXIT);
  });

  it("sets stage to TOUCH_SLIDING_ANIMATION_EXIT on touch end", () => {
    // mount
    component = mount(<ReactNotificationComponent />);

    // add notification
    let id = addNotification(getNotificationMock());

    // toggle touch end
    instance().toggleTouchEnd({ id });

    // find by id
    let item = state().notifications.find(elem => elem.id === id);

    // expect stage to be set
    expect(item).not.toBeUndefined();
    expect(item.stage).toBe(NOTIFICATION_STAGE.TOUCH_SLIDING_ANIMATION_EXIT);
  });

  it("toggles notification removal", () => {
    // mount
    component = mount(<ReactNotificationComponent />);

    // add notifications
    const id = addNotification(getNotificationMock());

    // expect lengt to match number of added notifications
    expect(state().notifications.length).toBe(1);

    // toggle removal
    instance().toggleRemoval({ id });

    // tick
    clock.tick(100);

    // expect decrease
    expect(component.state().notifications.length).toBe(0);
  });

  it("onNotificationRemoval is called on notification removal if defined", () => {
    const mock = jest.fn();

    // mount with callback prop `onNotificationRemoval`
    component = mount(<ReactNotificationComponent onNotificationRemoval={mock} />);

    // add notification
    const id = addNotification(getNotificationMock());

    // trigger removal
    instance().toggleRemoval({ id });

    expect(mock).toHaveBeenCalledTimes(1);
  });

  it("renders notification properly", () => {
    // mount
    component = mount(<ReactNotificationComponent />);

    // render notifications based on array input
    let res = instance().renderReactNotifications([getNotificationMock()]);

    // one notification rendered
    expect(res.length).toBe(1);

    let notif = res[0];
    expect(notif.props.notification).toBeDefined();
    expect(notif.props.isFirstNotification).toBeDefined();
    expect(notif.props.onClickHandler).toBeDefined();
    expect(notif.props.toggleRemoval).toBeDefined();
    expect(notif.props.toggleTimeoutRemoval).toBeDefined();
    expect(notif.props.toggleTouchEnd).toBeDefined();
  });

  it("renders mobile for 512px width if `isMobile` is true", () => {
    // width for mobile view
    global.window.innerWidth = 512;

    // just in case we update breakpoint in near future
    expect(global.window.innerWidth).toBeLessThan(BREAKPOINT);

    // mount
    component = mount(<ReactNotificationComponent />);

    // all mobile containers rendered
    expect(component.find(".notification-container-mobile-top").length).toBe(1);
    expect(component.find(".notification-container-mobile-bottom").length).toBe(1);
  });

  it("renders desktop for 512px width if `isMobile` is false", () => {
    // width for mobile view
    global.window.innerWidth = 512;

    // just in case we update breakpoint in near future
    expect(global.window.innerWidth).toBeLessThan(BREAKPOINT);

    // mount
    component = mount(<ReactNotificationComponent isMobile={false} />);

    // all desktop containers rendered
    expect(component.find(".notification-container-top-left").length).toBe(1);
    expect(component.find(".notification-container-top-right").length).toBe(1);
    expect(component.find(".notification-container-bottom-left").length).toBe(1);
    expect(component.find(".notification-container-bottom-right").length).toBe(1);
  });

  it("renders desktop view", () => {
    // width for desktop view
    global.window.innerWidth = 1024;

    // mount component
    component = mount(<ReactNotificationComponent />);

    // all desktop containers rendered
    expect(component.find(".notification-container-top-left").length).toBe(1);
    expect(component.find(".notification-container-top-right").length).toBe(1);
    expect(component.find(".notification-container-bottom-left").length).toBe(1);
    expect(component.find(".notification-container-bottom-right").length).toBe(1);

    // no mobile containers rendered
    expect(component.find(".notification-container-mobile-top").length).toBe(0);
    expect(component.find(".notification-container-mobile-bottom").length).toBe(0);
  });

  it("renders mobile view", () => {
    // width for mobile view
    global.window.innerWidth = 512;

    // mount component with isMobile prop
    component = mount(<ReactNotificationComponent isMobile={true} />);

    // expect mobile containers to be rendered
    expect(component.find(".notification-container-mobile-top").length).toBe(1);
    expect(component.find(".notification-container-mobile-top").length).toBe(1);

    // expect desktop containers not to be rendered
    expect(component.find(".notification-container-top-left").length).toBe(0);
    expect(component.find(".notification-container-top-right").length).toBe(0);
    expect(component.find(".notification-container-bottom-left").length).toBe(0);
    expect(component.find(".notification-container-bottom-right").length).toBe(0);
  });

  it("changes from desktop to mobile", () => {
    global.window.width = 1024;

    // mount component with isMobile prop
    component = mount(<ReactNotificationComponent isMobile={true} />);

    // manually update width to render mobile containers
    component.setState({ width: 512 });

    // expect mobile containers to be rendered
    expect(component.find(".notification-container-mobile-top").length).toBe(1);
    expect(component.find(".notification-container-mobile-bottom").length).toBe(1);

    // expect desktop containers not to be rendered
    expect(component.find(".notification-container-top-left").length).toBe(0);
    expect(component.find(".notification-container-top-right").length).toBe(0);
    expect(component.find(".notification-container-bottom-left").length).toBe(0);
    expect(component.find(".notification-container-bottom-right").length).toBe(0);
  });

  it("changes from mobile to desktop", () => {
    global.window.width = 512;

    // mount component
    component = mount(<ReactNotificationComponent isMobile={true} />);

    // set state to render mobile view
    component.setState({ width: 1024 });

    // expect no mobile containers to be rendered
    expect(component.find(".notification-container-mobile-top").length).toBe(0);
    expect(component.find(".notification-container-mobile-bottom").length).toBe(0);

    // expect desktop containers to be rendered
    expect(component.find(".notification-container-top-left").length).toBe(1);
    expect(component.find(".notification-container-top-right").length).toBe(1);
    expect(component.find(".notification-container-bottom-left").length).toBe(1);
    expect(component.find(".notification-container-bottom-right").length).toBe(1);
  });
});