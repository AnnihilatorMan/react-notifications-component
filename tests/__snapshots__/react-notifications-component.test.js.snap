// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wrapper component component does not set state to default values on initialization 1`] = `
Object {
  "breakpoint": 320,
  "isMobile": false,
  "notifications": Array [],
  "width": 1024,
}
`;

exports[`Wrapper component component sets state to default values on initialization 1`] = `
Object {
  "breakpoint": 768,
  "isMobile": true,
  "notifications": Array [],
  "width": 1024,
}
`;

exports[`Wrapper component component sets userDefinedTypes on initialization 1`] = `
Object {
  "breakpoint": 768,
  "isMobile": true,
  "notifications": Array [],
  "userDefinedTypes": Array [
    Object {
      "htmlClasses": Array [
        "awesome",
      ],
      "name": "awesome",
    },
  ],
  "width": 1024,
}
`;
